Populate NodeResourceTopology API

swatisehgal opened the pull request:
**What this PR does / why we need it**:
This API is needed for Topology aware scheduling work in Kubernetes and it was [suggested](https://github.com/kubernetes/kubernetes/pull/96275#discussion_r1119165747) that a CRD based API in staging is not ideal and it would be better to find a more suitable location under Kubernetes umbrella (under kubernetes-sigs). The repo was created under kubernetes-sigs (https://github.com/kubernetes/org/issues/4224) and this PR populates the repo.



**Background information/ Additional context**
[NFD Topology Updater](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/docs/get-started/introduction.md#nfd-topology-updater) in [Node feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery) and [Topology aware scheduler plugin](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/pkg/noderesourcetopology) in [scheduler-plugins](https://github.com/kubernetes-sigs/scheduler-plugins) are two components for enabling Topology aware scheduling in Kubernetes (additional details about these components can be found below). These two projects are based in [kubernetes-sigs](https://github.com/kubernetes-sigs) organisation and depend on [https://github.com/k8stopologyawareschedwg/noderesourcetopology-api](https://github.com/k8stopologyawareschedwg/noderesourcetopology-api ) which is not operated by Kubernetes github management team.

**Additional information about NFD Topology Updater and scheduler plugin**
1. NFD Topology Updater, a daemon [depends](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/go.mod#L12) on NodeResourceTopology API to expose resource hardware topology on a per NUMA basis via CRs that correspond to all the nodes in the cluster.
2. Topology aware scheduler plugin, an out of tree scheduler plugin [depends](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/go.mod#L8) on NodeResourceTopology API to consumed the CRs that corresponds to nodes (created by NFD Topology Updater) to make an intelligent NUMA aware scheduling decision.

**Why do we need Topology aware scheduling in Kubernetes?**
With the introduction of Topology Manager in Kubelet, pods could be scheduled on a node where total amount of resources available satisfy the resources requested, but resource distribution could not satisfy the appropriate Topology Manager policy leading to pod admission failure. Topology aware scheduler aims to empower scheduler to make intelligent topology aware placement decisions optimizing cluster wide performance of workloads.


**Which issue(s) this PR fixes**:
https://github.com/kubernetes/community/issues/6308
https://github.com/kubernetes/kubernetes/issues/84869



**Additional notes/documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- Node Feature Discovery Issue: [Exposing Hardware Topology through CRDs in Node feature discovery](https://github.com/kubernetes-sigs/node-feature-discovery/issues/333)
- Topology Aware Scheduling Use Cases: [here](https://docs.google.com/document/d/1OJAejm4kdRDUqgWdlYQj9b9DBcLmZY71PrLH7eSvX6I/edit#heading=h.fz4hmiaj33sv)
- Kubernetes Node Resource Topology Custom Resource Definition Standard: [here](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit)
- NFD 
- [Other doc]: Based on ideas that were discussed in following documents:
  * [Kubernetes Topology-Aware Scheduling](https://docs.google.com/document/d/1gPknVIOiu-c_fpLm53-jUAm-AGQQ8XC0hQYE7hq0L4c/edit?ts=5ea71ac0#)
  * [Ideas of making Scheduler to be NUMA-aware](https://docs.google.com/presentation/d/1u6fD_8xf40dMTbpwBQmanOpI-WqHioryovHozUDfKhU/edit#slide=id.g7fc7df652a_29_2)
  * [NUMA-aware scheduling 
](https://docs.google.com/presentation/d/1y2ObdZUtMFp2Z0EeyUu1cm469EQNMBjJbhjKiHwSeQo/edit#slide=id.p)

swatisehgal committed:
Add NRT API

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
Setup go modules

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
Autogenerated code

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
docs: Document and add details on API fields

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
Repo bootstrapping

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
manifest: Add manifests

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
owners: Add OWNERS_ALIASES file and add Kubernetes API reviewers

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/hold
to prevent inadvertent merge

ffromani commented on go.mod:
```diff
@@ -0,0 +1,54 @@
+
+require (
+	github.com/gogo/protobuf v1.3.2
+	k8s.io/apimachinery v0.22.3
+	k8s.io/client-go v0.22.3
+	k8s.io/code-generator v0.22.3
```
why so old versions?

swatisehgal replied:
I will bump the deps to newer version, Thanks for pointing this out.

swatisehgal replied:
I added additional commits but if squashing of commits is preferred, I can do that.

ffromani reviewed (commented):
non-binding LGTM, want to have another pass later
question inside

swatisehgal committed:
hack: Update to using kube_codegen.sh instead of generate-groups.sh

We are getting the following warnings if we continue using generate-groups.sh
so updating to kube_codegen.sh instead as per the recommendation:

```
WARNING: generate-groups.sh is deprecated.
WARNING: Please use k8s.io/code-generator/kube_codegen.sh instead.
```

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

marquiz commented on README.md:
```diff
@@ -1,25 +1,16 @@
-  [Embargo Policy](https://git.k8s.io/security/private-distributors-list.md#embargo-policy)
-  and will be removed and replaced if they violate that agreement.
+## Purpose
+This repository contains the CRD based API definition used for enabling NUMA aware Scheduling in Kubernetes.
```
Should we talk more generic "topology aware scheduling" instead of narrow "NUMA" scope?

marquiz commented on go.mod:
Run `go mod tidy` to update this

marquiz commented on pkg/apis/topology/v1alpha1/register.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: topology.GroupName, Version: "v1alpha2"}
```
alpha2??

swatisehgal replied:
Good catch, will fix.

marquiz commented on manifests/cr-instance-example.yaml:
```diff
@@ -0,0 +1,43 @@
+metadata:
+  name: node1
+attributes:
+  Name: "topologyManagerPolicy"
```
nit: should we follow lowercase (or camelcase) for the field names?
```suggestion
  name: "topologyManagerPolicy"
```

swatisehgal replied:
Sure, this is just an example. Typically the CR instance would be populated by the exporter component and is expected to have the fields in lowercase. I will update the example here.

marquiz reviewed (commented):
Thanks @swatisehgal for the PR. And apologies for the slow response time 🙄

Looks generally good, just a few small comments below.

swatisehgal commented:
/assign @liggitt @thockin
for API reviews based on the suggestion [here](https://github.com/kubernetes/org/issues/4224#issuecomment-1559961067).

vsoch commented:
hey folks! My team is interested in this work (and I could help if needed). Is there a status update / what are next steps?

swatisehgal commented:
> hey folks! My team is interested in this work (and I could help if needed). Is there a status update / what are next steps?

Hi @vsoch The current status is that we are waiting for this PR to be reviewed by Kubernetes API reviewers. I am conscious that we are close to the holiday season and a lot of people are already out  so will be actively pursuing soliciting of reviews from API reviewers in the beginning of new year.

If you have bandwidth and would like, you are more than welcome to review the PR and let us know your feeedback.

vsoch commented on API.md:
```diff
@@ -0,0 +1,93 @@
+
+1. NodeResourceTopology.Zones - List of Zones. A zone can represent NUMA nodes, packages, dies, cores etc.
+    1. Name - Name of a zone e.g numa-node-0, package-0
+	1. Type - Type of a zone e.g. NumaNode, Package
```
```suggestion
	1. Type - Type of a zone e.g. NUMANode, Package
```
Nit: NUMA is usually in all caps. If it's for a Kubernetes struct name (and that goes against convention) then what you have is probably OK!

swatisehgal replied:
Sure, can update this. These zone names are usually populated by the component using this API to expose instances. This is just an example of what a Zone can represent and can be named, in the examples below we provided examples we have Zone type specified as NUMA.

vsoch commented on API.md:
```diff
@@ -0,0 +1,93 @@
+
+# Definitions of Keys
+
+1. NodeResourceTopology.Zones - List of Zones. A zone can represent NUMA nodes, packages, dies, cores etc.
```
It would be good to have just `NodeResourceTopology` -> NRT somewhere before it's mentioned below. It's not impossible to figure out, but might as well be explicit.

swatisehgal replied:
Sure, will add that.

vsoch commented on API.md:
```diff
@@ -0,0 +1,93 @@
+	Name       string           `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
```
Will the idea here be that costs are compared with some Less (or similar function) to determine an optimal placement? I'm guessing cost is akin to an edge - is that correct? If this is a graph is there any reason to not model it more explicitly as one (naming-wise)?  And there are multiple in the list to allow for different kinds of costs (e.g., different edges - a node to node cost vs. node to something else?) Why was a list chosen instead of a map for easy lookup?

vsoch replied:
Is Zone the best term to describe a resource group? I know it has an explicit meaning for hardware, but in the context of cloud native folks we will think of something else. Is there a qualifier we can add?

marquiz replied:
> Will the idea here be that costs are compared with some Less (or similar function) to determine an optimal placement?

Yes, the cost basically represents the "topological distance" between zones. Or the cost of moving data between them.

> I'm guessing cost is akin to an edge - is that correct?

Yes. It can be thought as the weight of an edge in a weighed graph.

> If this is a graph is there any reason to not model it more explicitly as one (naming-wise)?

I think the naming was chosen for the context (hardware topology) instead of using just generic abstract naming of graphs. Any counterproposals?

> And there are multiple in the list to allow for different kinds of costs (e.g., different edges - a node to node cost vs. node to something else?)

It's not for different types of costs but cost to different zones. If you have 4 zones A, B, C and D you want cost between each of them, e.g. for A: A->B, A->C and A->D

> Why was a list chosen instead of a map for easy lookup?

I think one reason is the k8s API conventions/guidelines to prefer lists of named objects over maps.

> Is Zone the best term to describe a resource group? I know it has an explicit meaning for hardware, but in the context of cloud native folks we will think of something else. Is there a qualifier we can add?

It's not supposed to be just some "random" resource group. The assumption kinda is that it's some topologically relevant group. So I think zone is good but I'm open to hear other opinions.

Anything to correct or add @swatisehgal @ffromani ?

ffromani replied:
Agree with everything and still on the lookout to find a reference that k8s APIs prefer lists over maps, because I also clearly remember that suggestions but can't find it atm.

swatisehgal replied:
Agree with everything here too. Thanks @marquiz!

> > Why was a list chosen instead of a map for easy lookup?
> 
> I think one reason is the k8s API conventions/guidelines to prefer lists of named objects over maps.

Yep, that's correct. [Here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps) is the reference to API conventions because of which we updated this API to use lists instead of maps.
 
> > Is Zone the best term to describe a resource group? I know it has an explicit meaning for hardware, but in the context of cloud native folks we will think of something else. Is there a qualifier we can add?
> 
> It's not supposed to be just some "random" resource group. The assumption kinda is that it's some topologically relevant group. So I think zone is good but I'm open to hear other opinions.

Agreed, we were aiming for an umbrella term that can be used to represent a range of topological zones in hardware and the term Zone seemed to fit the bill. From cloud native context perspective, if you have better naming suggestions we would be happy to consider.

vsoch replied:
> range of topological zones in hardware 

Do you have a document or reference that talks about hardware zones? I'm new to the idea (and want to brainstorm a possible other term, but perhaps there is not a good replacement)!

swatisehgal replied:
The idea is that a zone could be any topological zone like core, NUMANode, package, die, etc. We have examples in [this](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit?usp=sharing) document that might help.

vsoch reviewed (commented):
A few comments - still reviewing!

KunWuLuan commented on API.md:
```diff
@@ -0,0 +1,93 @@
+	Name       string           `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
```
Is cost for a zone enough for resources? Cost of cpus in different NUMA nodes can be different with gpus in different NUMA nodes because of the NVLink or NVSwitch.

marquiz replied:
I think resources with special interconnects between them should then be modeled as separate zones, if you want to represent that in detail

ffromani replied:
> I think resources with special interconnects between them should then be modeled as separate zones, if you want to represent that in detail

+1 from my side, perhaps we should "just" note in the docs (= API comments) that this is how envision the API to be used

swatisehgal replied:
Good idea, will try to capture this in the docs.

KunWuLuan replied:
Do you mean we can have a zoneList like this?
{{NUMANode-gpu-0,cost=10}, {PCIE-0-gpu1, type=gpu, parent=NUMANode-gpu-0},{PCIE-0-gpu2, type=gpu, parent=NUMANode-gpu-0},{NUMANode-cpu-0,cost=20},{PCIE-0-cpu1, type=cpu, parent=NUMANode-cpu-0},{PCIE-0-cpu2, type=cpu, parent=NUMANode-cpu-0},}

swatisehgal replied:
Yes, this is close but the thing to note here is that the cost is a list that represents distance of that zone from other zones and we represent this with a key-value pair (key capturing the zone and value capturing the cost) so in the example above cost should be indicative of distance of NUMANode-gpu-0 from NUMANode-cpu-0 in both the cases and therefore should be identical. But if there was another zone in the example we could have a different cost from these zones. Something like below:

```
{{NUMANode-gpu-0,cost=({NUMANode-cpu-0:10},{NUMANode-gpu-1:20}}}, {PCIE-0-gpu1, type=gpu, parent=NUMANode-gpu-0},{PCIE-0-gpu2, type=gpu, parent=NUMANode-gpu-0},
{NUMANode-cpu-0,cost=({NUMANode-gpu-0:10},{NUMANode-gpu-1:30}},{PCIE-0-cpu1, type=cpu, parent=NUMANode-cpu-0},{PCIE-0-cpu2, type=cpu, parent=NUMANode-cpu-0},
{NUMANode-gpu-1,cost=({NUMANode-cpu-0:30},{NUMANode-gpu-0:20}}, {PCIE-1-gpu1, type=gpu, parent=NUMANode-gpu-1},{PCIE-1-gpu2, type=gpu, parent=NUMANode-gpu-1}}
```

Hope this helps.

ffromani replied:
> Good idea, will try to capture this in the docs.

thanks! also providing examples of topologies and their modellization would help (obviously we can all contribute examples).
I guess we will also need  somehow to address the point that there are multiple legal and equivalent modelization of the same hardware, not just one. Unlikely a blocker for alpha, though.

KunWuLuan replied:
Thanks for replying. I have another question. When I tried to write a sample for the case, I found that we have to add type in zoneName to avoid create same name for different type of device in one node or other level, do we still need a "type" property? Maybe type can be an attribute. HDYT? @swatisehgal

KunWuLuan replied:
And if I want to define a gpu in socket 0, numa 0, pcie 0. I have to define a list with 3 items. 
Maybe I want to write in this format: 
```
{Name: gpu-0, Attributes={Socket=0, Numa=0, PCIE=0}}
```
In what cases we need to care about the parent of differnet zones?

swatisehgal replied:
> And if I want to define a gpu in socket 0, numa 0, pcie 0. I have to define a list with 3 items. Maybe I want to write in this format:
> 
> ```
> {Name: gpu-0, Attributes={Socket=0, Numa=0, PCIE=0}}
> ```


We were aiming for an API that provides flexibility and has a hierarchical structure to represent relationship and proximity between various topological zones. It sounds like you were expecting a more flatter representation which is achievable with the current API we have. You can define all the resources within a single zone and use the attributes field like you are suggesting (assuming we add resource level attributes like we previously discussed).

It is important to keep [this](https://github.com/kubernetes-sigs/noderesourcetopology-api/pull/1#discussion_r1453326813) comment from @ffromani in mind. We wanted to represent a wide range of hardware and resources and can only provide our recommendations on how the API can be _potentially_ used. But, there is no right and wrong here, if you would like it to use differently you can totally do that!

As long as the producer (exporter that creates instances of object) and the consumer (which could be a human or a software entity) have a shared understanding of the representation we should be okay. 

> In what cases we need to care about the parent of different zones?

A simple example would to be represent hardware differently where there are multiple NUMA nodes per socket vs multiple sockets per NUMA nodes. I would suggest that you refer to [this](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit?usp=sharing) document for the rationale behind the design of this API.

swatisehgal replied:
> Thanks for replying. I have another question. When I tried to write a sample for the case, I found that we have to add type in zoneName to avoid create same name for different type of device in one node or other level, do we still need a "type" property? Maybe type can be an attribute. HDYT? @swatisehgal 

For the cases we have in mind, we need the "type" property as the zone type is typically something we want to define whereas attributes can be some additional zone specific attributes that the exporter might want to populate. It sounds like you would like it to be an optional field? Would making the type field optional help you for the way you would like to populate data?

KunWuLuan commented on API.md:
```diff
@@ -0,0 +1,93 @@
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
+	Attributes AttributeList    `json:"attributes,omitempty" protobuf:"bytes,5,rep,name=attributes"`
```
If there are more than one device on a NUMA node and these device can be different for some attributes, how should we set attributes for these devices?

marquiz replied:
These attributes are for the zone and cannot represent attributes that differ per-resource. For that, we'd need to extend the `ResourceInfo` type, I think.

@swatisehgal @ffromani

swatisehgal replied:
For resource level attributes, extending ResourceInfo is certainly one way. This seems like a reasonable addition to me.

Another way could be to create a new zone that is child of a zone representing a NUMA node and then capture these resources  with their corresponding attributes under the newly created sub-NUMA zone.

The former requires some changes while the latter doesn't. I don't feel too strongly about either so would like to hear @ffromani and @marquiz opinion on this.

marquiz replied:
> Another way could be to create a new zone that is child of a zone representing a NUMA node and then capture these resources with their corresponding attributes under the newly created sub-NUMA zone.

True, this should work.
 
> The former requires some changes while the latter doesn't. I don't feel too strongly about either so would like to hear @ffromani and @marquiz opinion on this.

I don't have strong opinions either. Resorting to the latter (no changes) until we understand the usage scenarios for the former(?)

swatisehgal commented:
@liggitt A gentle ping for api reviews.

liggitt commented:
will route this one to @thockin, it is much more aligned with the DRA reviews he's been doing and he has the topology stuff much more in his head

Present in https://github.com/orgs/kubernetes/projects/169/views/3?sliceBy%5Bvalue%5D=thockin

swatisehgal commented:
> will route this one to @thockin, it is much more aligned with the DRA reviews he's been doing and he has the topology stuff much more in his head
> 
> Present in https://github.com/orgs/kubernetes/projects/169/views/3?sliceBy%5Bvalue%5D=thockin

Thanks!

thockin commented:
Hi Swati,

I am about to get on a plane for KubeCon - is this OK to wait until after?

On Thu, Mar 14, 2024 at 9:06 AM Swati Sehgal ***@***.***>
wrote:

> will route this one to @thockin <https://github.com/thockin>, it is much
> more aligned with the DRA reviews he's been doing and he has the topology
> stuff much more in his head
>
> Present in
> https://github.com/orgs/kubernetes/projects/169/views/3?sliceBy%5Bvalue%5D=thockin
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/noderesourcetopology-api/pull/1#issuecomment-1997807786>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVCAQ6EEPYPKUUERT3TYYHDJBAVCNFSM6AAAAAA57EZG5SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJXHAYDONZYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vsoch commented:
He's leaaaavin' on a jet plane... :airplane:

swatisehgal commented:
> Hi Swati, I am about to get on a plane for KubeCon - is this OK to wait until after?

Hi Tim, sure no problem. Have a great KubeCon!

swatisehgal commented:
@thockin A gentle ping to review this PR. Thanks!

pacoxu commented on manifests/cr-instance-example.yaml:
```diff
@@ -0,0 +1,43 @@
+    allocatable:  51
+    available:    51
+    capacity:     52
+  name:     node-1
```
```
-  name:     node-1
```
Same to line 12.

SergeyKanzhelev commented:
/assign @johnbelamaric

SergeyKanzhelev commented:
/remove-lifecycle stale
